### PR TITLE
 refactor(provider/kubernetes): Remove reflection from cache constructors

### PR DESCRIPTION
**Pull request by ezimanyi**
* refactor(provider/kubernetes): Remove reflection from cache constructors 

  Constructing caching agents currently uses reflection, which breaks static analysis and makes it harder to reason about and refactor this code. Instead of having each handler return a reference to the class that should be constructed, instead have the handler return a functional interface to the actual constructor so that the base class does not need to use reflection.

* refactor(kubernetes): Runtime -> compile time error 

  All subclasses of KubernetesV2Caching agent need to implement either primaryKinds or primaryKind, and a failure to do so is a runtime error.  Make this a compile-time error by standardizing on primaryKinds and making it abstract in the base class.

### CODE REVIEW

**Comment by ezimanyi** [comment]
@plumpy : Realized I can't formally request review from you yet, but this is the PR I just mentioned.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/op/handler/KubernetesHandler.java [hunk 1]
```diff
@@ -96,43 +88,15 @@ public KubernetesV2CachingAgent buildCachingAgent(
       int agentCount,
       Long agentInterval
   ) {
-    Constructor constructor;
-    Class<? extends KubernetesV2CachingAgent> clazz = cachingAgentClass();
```
Comment (jtk54) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/op/handler/KubernetesHandler.java]: Always feels good to get rid of variables named `clazz`.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/caching/agent/KubernetesV2CachingAgentFactory.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+package com.netflix.spinnaker.clouddriver.kubernetes.v2.caching.agent;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.netflix.spectator.api.Registry;
+import com.netflix.spinnaker.clouddriver.kubernetes.security.KubernetesNamedAccountCredentials;
+import com.netflix.spinnaker.clouddriver.kubernetes.v2.description.KubernetesResourcePropertyRegistry;
+import com.netflix.spinnaker.clouddriver.kubernetes.v2.security.KubernetesV2Credentials;
+
+public interface KubernetesV2CachingAgentFactory {
```
Comment (plumpy) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/caching/agent/KubernetesV2CachingAgentFactory.java]: Maybe mark this @FunctionalInterface?

Comment (ezimanyi) [reply clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/caching/agent/KubernetesV2CachingAgentFactory.java]: Good idea, will do that

**Pull request merged**